SdmxDataParser/src/main/java/org/sdmxsource/sdmx/dataparser/engine/writer/streaming/StreamDataWriterEngineBase.java

```diff
@@ -656,9 +657,10 @@ private void writeMessageTag() throws XMLStreamException {
      * Writes out the dataset header to the writer
      *
      * @param dataSetInfo contains base information about dataset
+     * @param datasetAction contains action performed on dataset
      * @throws XMLStreamException the xml stream exception
      */
-    protected void writeDatasetHeader(DataSetInfo dataSetInfo) throws XMLStreamException {
+    protected void writeDatasetHeader(DataSetInfo dataSetInfo, DATASET_ACTION datasetAction) throws XMLStreamException {
```

Comment:
change of the protected method (which might be extended by other clients) is a bit risky, since this is a breaking change
I would recommend trying to just re-initialise dataSetInfo object, on **line 334**, like this:
```
dataSetInfo = new DataSetInfo(dataSetInfo.provision, dataSetInfo.dataflow, dataSetInfo.dsd, header);
```
so that action which resides in header will be updated, but without the need to expose implementation classes.